src/utils/cloudflare/api.js

```diff
@@ -0,0 +1,114 @@
+const assert = require('assert');
+const retry = require('promise-retry');
+const { helpers: { generateClass } } = require('common-errors');
+
+const { CloudflareClient } = require('./client');
+
+const RULE_LIST = 'accounts/:accountId/rules/lists';
+const RULE_ITEM_LIST = 'accounts/:accountId/rules/lists/:listId/items';
+const RULE_BATCH_PROGESS = 'accounts/:accountId/rules/lists/bulk_operations/:operationId';
+
+const BulkOperationError = generateClass('BulkOperationError', {
+  args: ['message', 'response'],
+});
+
+const defaultRetryConfig = {
+  retries: 10,
+  factor: 0.5,
+};
+
+class CloudflareAPI {
+  constructor(cfClient, config = {}, withProps = {}) {
+    assert(cfClient instanceof CloudflareClient, 'invalid client');
+    this.config = config;
+    this.cfClient = cfClient;
+    this.http = cfClient.getHttpClient();
+    this.props = withProps;
+  }
+
+  withAccountId(accountId) {
+    return new CloudflareAPI(this.cfClient, this.config, { accountId });
+  }
+
+  apiUrl(t, values) {
+    const { accountId } = this.props;
+    assert(accountId, 'accountId invalid');
+    return Object.entries({ accountId, ...values })
+      .reduce((prev, [prop, val]) => prev.replace(`:${prop}`, val), t);
+  }
+
+  async createList(name, kind = 'ip', description = '') {
+    const { accountId } = this.props;
+    const url = this.apiUrl(RULE_LIST, { accountId });
+    const { result } = await this.http.post(url, {
+      json: { name, kind, description },
+    });
+    return result;
+  }
+
+  async getLists() {
+    const { accountId } = this.props;
+    return this.http.get(this.apiUrl(RULE_LIST, { accountId }));
+  }
+
+  async getListItems(listId, cursor) {
+    const { accountId } = this.props;
+    const url = this.apiUrl(RULE_ITEM_LIST, { accountId, listId });
+    return this.http.get(url, { searchParams: { cursor } });
+  }
+
+  async waitListOperation(operationId) {
+    const { accountId } = this.props;
+    const url = this.apiUrl(RULE_BATCH_PROGESS, { accountId, operationId });
+
+    // https://api.cloudflare.com/#rules-lists-get-bulk-operation
+    return retry(
+      async (retryFn) => {
+        try {
+          const response = await this.http.get(url);
```

Comment:
this implementation doesnt look good, try this module with retry by default and explicit error that needs to be thrown (pRetry.abort) to abort retries https://www.npmjs.com/package/p-retry

Reply:
Changed package + added extra tests